### ISSUE TITLE
[RFC] Make it possible to get and set files content

### ISSUE THREAD

**Issue by smortex** [choria-plugins/filemgr-agent#4]
Hi!

I am thinking about adding support for new operations for this agent, basically to read or write files and by extension set files owner / group / mode.

The purpose is to fill-in a gap I faced a few times for some very simple tasks that do not really deserve a new module to handle:

* Checking that the content of a file match some expectation (e.g. ensure a given setting is set properly in `postfix.conf`). This would be handy when the checksum of the file cannot be used to determine if something is write or wrong.  Matching the file against a regexp may be possible but feels overkill for this "simple" agent, so I would prefer to stick to the raw value, and letting the user `grep(1)` on the result if he wants to;
* Syncing (pushing) my local `~/.zshrc` to `/home/romain/.zshrc` on all nodes I work with, with proper user, group and permissions.

Before starting hacking, I would appreciate a green light / comments.  And since finding names is the hardest part of computing, suggestions for action names would be appreciated :wink:  

Thanks!

**Comment by ripienaar** [comment]
Hmm yeah though NATS and middleware in general are pretty bad for transferring files. Now sure your zsh file is tiny and no problem but the moment it exist people will try to copy all sorts of stuff - that’s my worry. And when it doesn’t work there is a feature that’s effectively guaranteed to fail. 

For regex matches and so that would be a good feature yes :)

**Comment by smortex** [comment]
> but the moment it exist people will try to copy all sorts of stuff - that’s my worry. And when it doesn’t work there is a feature that’s effectively guaranteed to fail.

Yeah, I was thinking about people trying to send stupid things such as ISO images, databases and the like.  I have not tested this and expect this to not be reliable :smile: …  What about trying to implement this "stupidly" and see what happen (attempt to measure the failures), and then setup some adequate fool-proof check that performs validation *before* data is transmitted?  Or do you think that it's probably safer to do this development for our code base only and not share it with the filemgr plugin?

> For regex matches and so that would be a good feature yes :)

Cool, that could look like this:

```sh-session
user@client $ mco filemgr match -f /etc/postfix/main.cf '^relayhost *= *[[:print:]]+$'
  node1.example.com: true
  node2.example.com: true
  node3.example.com: false
```

**Comment by ripienaar** [comment]
The broker keepsnthenentire file in memory and while transferring it which could take ages it will block other messages. Big messages are simply bad that’s all there is to it :)

We need an augeus agent again :)

Do you know data plugins? They don’t work yet with the. We server but they are made for this and the filemgr already has one that let you do finds of nodes matching things but yeah I like what you are thinking with the regex stuff

**Comment by smortex** [comment]
Ah, I was not aware that a large message would block other messages during it's transfer :-/

I have not played much with data plugins, I guess you mean something like `-S "match('/path/to/file', '^relayhost =').found = true"` or `-S "match('/path/to/file', '^relayhost = .*$').match = 'relayhost = mail.example.com'"`?  This totally makes sense.

I can't find the data plugin for filemgr, but I have a few other and they seem to be working as expected with he new server (I use `puppet()` one from time to time).  Or am I confused with what is "new" :-D

**Comment by ripienaar** [comment]
Hmm I could have sworn it used to have a data plugin but you are right I don’t see it now have to go look what happened. 

The go based server doesn’t yet support data plugins.
